tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,190 @@
+
+# zarr spec: https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html
+
+
+class SyncMemStoreWrapper(StoreABC):
+    def __init__(self, store_type):
+        """Store to hold async functions that map to StoreABC abstract methods"""
+        if store_type in ("Memory", "memory"):
+            self.store = MemoryStore(mode="w")  # store_container.store
+        elif store_type in ("Local", "local"):
+            self.store = LocalStore(mode="w")
+        elif store_type in ("Remote", "remote"):
+            self.store = RemoteStore(mode="w")
+        else:
+            raise ValueError(f"Invalid item_type: {store_type}")
+        # Unfortunately, hypothesis' stateful testing infra does not support asyncio
+
+    # So we redefine sync versions of the Store API.
+    # https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+    async def set(self, key, data_buffer):  # buffer is value
+        await self.store.set(key, data_buffer)
+
+    async def list(self):
+        paths = [path async for path in self.store.list()]
+        # note(f'(store set) store {paths=}, type {type(paths)=}, {len(paths)=}')
+        return paths
+
+    async def get(
+        self,
+        key,
+    ):
+        obs = await self.store.get(key, prototype=default_buffer_prototype())
+        return obs
+
+    async def get_partial_values(self, key_ranges):
+        obs_maybe = await self.store.get_partial_values(
+            prototype=default_buffer_prototype(), key_ranges=key_ranges
+        )
+        return obs_maybe
+
+    async def delete(self, path):  # path is key
+        await self.store.delete(path)
+
+    async def empty(self):
+        await self.store.empty()
+
+    async def clear(self):
+        await self.store.clear()
+
+    async def exists(self, key):
+        raise NotImplementedError
+
+    async def list_dir(self, prefix):
+        raise NotImplementedError
+
+    async def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    async def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    async def supports_listing(self):
+        raise NotImplementedError
+
+    async def supports_partial_writes(self):
+        raise NotImplementedError
+
+    async def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    def __init__(self):  # look into using run_machine_as_test()
+        super().__init__()
+        self.model = {}
+        self.sync_wrapper = SyncMemStoreWrapper("memory")
+        self.store = self.sync_wrapper.store
+        # self.store = MemoryStore(mode='w')
+
+    @rule(key=paths, data=st.binary(min_size=0, max_size=100))
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        asyncio.run(self.sync_wrapper.set(key, data_buf))
```

Comment:
```suggestion
        self.store.set(key, data_buf))
```

This is what we want to enable.